integrations/llama_cpp/src/haystack_integrations/components/generators/llama_cpp/chat/chat_generator.py

```diff
@@ -254,9 +304,95 @@ def run(
+
+        for i, chunk in enumerate(response_stream):
+            content = ""
+            tool_calls = []
+            mapped_finish_reason = None
+
+            # Track new tool call IDs in this chunk.
+            # Considering tool call ID is the only reliable way to recognize tool calls in llama.cpp streaming.
+            # They are often spread across multiple chunks.
+            new_tool_call_ids = set()
+
+            if chunk.get("choices"):
+                choice = chunk["choices"][0]
+                delta = choice.get("delta", {})
+
+                finish_reason = choice.get("finish_reason")
+                mapped_finish_reason = FINISH_REASON_MAPPING.get(finish_reason or "") if finish_reason else None
+
+                if content_raw := delta.get("content"):
+                    content = str(content_raw)
+
+                tool_calls_data = delta.get("tool_calls")
+                if tool_calls_data is not None and isinstance(tool_calls_data, list):
+                    for tool_call_chunk in tool_calls_data:
+                        tool_call_id = tool_call_chunk.get("id")
+                        is_new_tool_call = tool_call_id and tool_call_id not in seen_tool_call_ids
+
+                        if is_new_tool_call:
+                            new_tool_call_ids.add(tool_call_id)
+                            seen_tool_call_ids.add(tool_call_id)
+
+                        function_data = tool_call_chunk.get("function", {})
+
+                        # Only include tool_name if this is a new tool call
+                        tool_name = function_data.get("name", "") if is_new_tool_call else ""
+
+                        tool_calls.append(
+                            ToolCallDelta(
+                                index=tool_call_chunk.get("index"),
+                                id=tool_call_id,
+                                tool_name=tool_name,
+                                arguments=function_data.get("arguments"),
+                            )
+                        )
+
+            # start is True if it's the first chunk or if we have new tool call IDs
+            start = i == 0 or len(new_tool_call_ids) > 0
+
+            streaming_chunk = StreamingChunk(
+                content="" if tool_calls else content,  # prioritize tool calls over content when both are present
+                tool_calls=tool_calls,
+                component_info=component_info,
+                index=i,
+                start=start,
+                finish_reason=mapped_finish_reason,
+                meta={
+                    "model": chunk["model"],
+                    "received_at": datetime.fromtimestamp(chunk["created"], tz=timezone.utc).isoformat(),
+                },  # llama.cpp does not provide usage metadata during streaming
+            )
+
+            streaming_chunks.append(streaming_chunk)
+
+            # Stream the chunk
+            streaming_callback(streaming_chunk)
```

Comment:
One thing I've always wondered (in other Generators too): if `streaming_callback(streaming_chunk)` fail with an exception, it will probably break the streaming I guess. Has this ever happened? WDYT about wrapping this in try/except?

Reply:
Yes. It was never reported but can happen.
I added a try/except block.